charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py

```diff
@@ -82,44 +130,109 @@ def __init__(self, message):
-        self.charm = charm
+    """Base class that represents a requirer relation end.
+
+    Args:
+        requirer_charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        requirer_charm (CharmBase): variable for storing the requirer application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, requirer_charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(requirer_charm, relation_name)
+        # TODO: remove this attribute, we are not really using it at all.
+        # Leaving it to keep backwards compatibility.
+        self.requirer_charm = requirer_charm
         self.relation_name = relation_name
 
-    def get_relation_data(self):
-        if not self.model.unit.is_leader():
-            return
-        gateway = self.model.relations[self.relation_name]
-        if len(gateway) == 0:
+    @staticmethod
+    def _relation_preflight_checks(relation: Relation) -> None:
+        """Series of checks for the relation and relation data.
+        Args:
+            relation (Relation): the relation object to run the checks on
+        Raises:
+            GatewayRelationDataMissingError: if data is missing or incomplete
+            GatewayRelationMissingError: if there is no related application
+        """
+        # Raise if there is no related applicaton
+        if not relation:
             raise GatewayRelationMissingError()
-        if len(gateway) > 1:
-            raise GatewayRelationTooManyError()
 
-        remote_app = [
-            app
-            for app in gateway[0].data.keys()
-            if isinstance(app, Application) and not app._is_our_app
-        ][0]
+        # Extract remote app information from relation
+        remote_app = relation.app
+        # Get relation data from remote app
+        relation_data = relation.data[remote_app]
+
+        # Raise if there is no data found in the relation data bag
+        if not relation_data:
+            raise GatewayRelationDataMissingError("No data found in relation data bag.")
+
+        # Check if the relation data contains the expected attributes
+        expected_attributes = ["gateway_name", "gateway_namespace"]
+        missing_attributes = [
+            attribute for attribute in expected_attributes if attribute not in relation_data
+        ]
+        if missing_attributes:
+            raise GatewayRelationDataMissingError(f"Missing attributes: {missing_attributes}")
+
+    def get_relation_data(self) -> dict:
+        """Returns a dictionary with the Gateway information.
+
+        Raises:
+            GatewayRelationDataMissingError: if data is missing entirely or some attributes
+            GatewayRelationMissingError: if there is no related application
+        """
+        # Run pre-flight checks
+        relation = self.model.get_relation(self.relation_name)
```

Comment:
We count on `get_relation` to validate that we have <=1 related application.  That feels like something that could be easily missed - maybe we can add a comment here?
```suggestion
        # Raises TooManyRelatedAppsError if related to more than one app
        relation = self.model.get_relation(self.relation_name)
```